ddtrace/llmobs/_evaluators/ragas/faithfulness.py

```diff
@@ -13,24 +71,242 @@ class RagasFaithfulnessEvaluator:
+
+        with self.llmobs_service.workflow(
+            "ragas.faithfulness", ml_app=_get_ml_app_for_ragas_trace(span_event)
+        ) as ragas_faithfulness_workflow:
+            try:
+                faithfulness_inputs = self._extract_faithfulness_inputs(span_event)
+                if faithfulness_inputs is None:
+                    logger.debug(
+                        "Failed to extract question and context from span sampled for ragas_faithfulness evaluation"
+                    )
+                    return None
+
+                question = faithfulness_inputs["question"]
+                answer = faithfulness_inputs["answer"]
+                context = faithfulness_inputs["context"]
+
+                statements = self._create_statements(question, answer)
+                if statements is None:
+                    logger.debug("Failed to create statements from answer for `ragas_faithfulness` evaluator")
+                    return None
+
+                faithfulness_list = self._create_verdicts(context, statements)
+                if faithfulness_list is None:
+                    logger.debug("Failed to create faithfulness list `ragas_faithfulness` evaluator")
+                    return None
+
+                score = self._compute_score(faithfulness_list)
+                if math.isnan(score):
+                    logger.debug("Score computation returned NaN for `ragas_faithfulness` evaluator")
+                    return None
+
+                return score
+            finally:
+                self.llmobs_service.annotate(
+                    span=ragas_faithfulness_workflow,
+                    input_data=span_event,
+                    output_data=score,
+                    metadata={
+                        "statements": statements,
+                        "faithfulness_list": faithfulness_list.dicts() if faithfulness_list else None,
+                    },
+                )
+
+    def _create_statements(self, question: str, answer: str) -> Optional[List[str]]:
+        with self.llmobs_service.workflow("ragas.create_statements"):
+            statements_prompt = self._create_statements_prompt(answer=answer, question=question)
+
+            """LLM step to break down the answer into simpler statements"""
+            statements = self.ragas_faithfulness_instance.llm.generate_text(statements_prompt)
+
+            statements = self.llm_output_parser_for_generated_statements.parse(statements.generations[0][0].text)
+
+            if statements is None:
+                return None
+            statements = [item["simpler_statements"] for item in statements.dicts()]
+            statements = [item for sublist in statements for item in sublist]
+
+            if not isinstance(statements, List):
+                return None
+            return statements
+
+    def _create_verdicts(self, context: str, statements: List[str]) -> Optional["StatementFaithfulnessAnswers"]:
```

Comment:
Does this work when you add a string into the `Optional[]` hint?

Reply:
Yep, we use stringified type hints here since StatementFaithfulnessAnswers might not exist if ragas deps aren't present